sklearn/decomposition/_dict_learning.py

```diff
@@ -385,54 +387,40 @@ def _update_dict(dictionary, Y, code, verbose=False, return_r2=False,
         Whether to enforce positivity when finding the dictionary.
 
         .. versionadded:: 0.20
-
-    Returns
-    -------
-    dictionary : ndarray of shape (n_features, n_components)
-        Updated dictionary.
     """
-    n_components = len(code)
-    n_features = Y.shape[0]
+    n_samples, n_components = code.shape
     random_state = check_random_state(random_state)
-    # Get BLAS functions
-    gemm, = linalg.get_blas_funcs(('gemm',), (dictionary, code, Y))
-    ger, = linalg.get_blas_funcs(('ger',), (dictionary, code))
-    nrm2, = linalg.get_blas_funcs(('nrm2',), (dictionary,))
-    # Residuals, computed with BLAS for speed and efficiency
-    # R <- -1.0 * U * V^T + 1.0 * Y
-    # Outputs R as Fortran array for efficiency
-    R = gemm(-1.0, dictionary, code, 1.0, Y)
+
+    if A is None:
+        A = np.dot(code.T, code)
+    if B is None:
+        B = np.dot(Y.T, code)
+
     for k in range(n_components):
-        # R <- 1.0 * U_k * V_k^T + R
-        R = ger(1.0, dictionary[:, k], code[k, :], a=R, overwrite_a=True)
-        dictionary[:, k] = np.dot(R, code[k, :])
-        if positive:
-            np.clip(dictionary[:, k], 0, None, out=dictionary[:, k])
-        # Scale k'th atom
-        # (U_k * U_k) ** 0.5
-        atom_norm = nrm2(dictionary[:, k])
-        if atom_norm < 1e-10:
+        if A[k, k] > 1e-6:
+            dictionary[k] += (B[:, k] - A[k].dot(dictionary)) / A[k, k]
+        else:
+            # kth atom is almost never used -> sample a new one from the data
             if verbose == 1:
                 sys.stdout.write("+")
                 sys.stdout.flush()
             elif verbose:
                 print("Adding new random atom")
-            dictionary[:, k] = random_state.randn(n_features)
-            if positive:
-                np.clip(dictionary[:, k], 0, None, out=dictionary[:, k])
-            # Setting corresponding coefs to 0
-            code[k, :] = 0.0
-            # (U_k * U_k) ** 0.5
-            atom_norm = nrm2(dictionary[:, k])
-            dictionary[:, k] /= atom_norm
-        else:
-            dictionary[:, k] /= atom_norm
-            # R <- -1.0 * U_k * V_k^T + R
-            R = ger(-1.0, dictionary[:, k], code[k, :], a=R, overwrite_a=True)
-    if return_r2:
-        R = nrm2(R) ** 2.0
-        return dictionary, R
-    return dictionary
+
+            newd = Y[random_state.choice(n_samples)]
+
+            # add small noise to avoid making the sparse coding ill conditioned
```

Comment:
Sampling from the data without adding noise triggered some warnings in the sparse coding at some point about ill conditioned problem. Adding a small noise fixed it. Do you have an opinion about that strategy @tomMoral or @agramfort ?

Reply:
Yes it is usually a good idea to add some noise. Else, you have a chance that you sample an atom that is only used to encode one variable. Adding some noise makes it somewhat more robust.

Note that instead of sampling the data, you could also sample from the residuals. These approaches are in particular appealing because they are related to the greedy algorithms that have somewhat some convergence guarantees.

Reply:
@ogrisel The warning came from here https://github.com/scikit-learn/scikit-learn/blob/266a11b2e17cf86effefe7b498b61ca31217ad31/sklearn/linear_model/_least_angle.py#L605

Reply:
@jeremiedbb what do you think of @tomMoral's comment above? The current PR is already a big net improvement so maybe it's not necessary. But maybe it would converge even faster or to a en even better value of the objective function.

Reply:
But if it makes the code more complex to organize, maybe this could be explored in a later PR to keep this PR simple to review and quick to merge.

Reply:
I was really satisfied by the strategy I implemented in the sense that now all atoms are used after only a few iterations.

Sampling from the residuals should give very good results too but it requires keeping track of the residuals. The previous implementation was keeping track of the residuals but was much slower and very hard to follow.